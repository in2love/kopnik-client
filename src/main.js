import Vue from 'vue'
import vuetify from './plugins/vuetify'
import router from './plugins/vue-router'
import i18n from './plugins/i18n'
import "./plugins/vee-validate"

import {Icon} from 'leaflet'
import 'leaflet/dist/leaflet.css'
// this part resolve an issue where the markers would not appear
delete Icon.Default.prototype._getIconUrl;
Icon.Default.mergeOptions({
    iconRetinaUrl: require('leaflet/dist/images/marker-icon-2x.png'),
    iconUrl: require('leaflet/dist/images/marker-icon.png'),
    shadowUrl: require('leaflet/dist/images/marker-shadow.png')
})

import AppVue from './components/AppVue.vue'
import {container} from "./plugins/bottle"
import './plugins/className'

Vue.config.productionTip = false

global.application = container.application
global.application.authenticate()
/*    .then(() => {
        if (!this.user) {
            this.section = 'Map'
        } else if (this.user.status == Kopnik.Status.NEW || this.user.status == Kopnik.Status.DECLINED) {
            this.section = 'Profile'
        } else {
            this.section = 'Map'
        }
    })*/

global.vue = new Vue({
    ...AppVue,
    el: "#appContainer",
    vuetify,
    i18n,
    router,
    // render: h => h(App)
})//.$mount('#app')
