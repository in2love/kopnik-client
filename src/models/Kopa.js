import {sync, collection, scalar, object} from '../decorators/sync'
import AbstractSync from "./AbstractSync";


export default class Kopnik extends AbstractSync {
    @scalar id = undefined

    @scalar theme = undefined
    /**
     *
     * @type {Kopnik}
     */
    @object initiator = undefined
    /**
     *
     * @type {Kopnik[]}
     */
    @collection parts = []

    stupidAdd(who) {
        this.parts.push(who)
    }

    /**
     * Пригласить копника
     *
     * @param {Kopnik} who
     */
    add(who) {
        if (this.isAdded(who)) {
            return
        }
        if (this.parts.length < 10) {
            this.parts.push(who)
        }
    }

    /**
     * не звать на копу
     * @param {Kopnik} who
     */
    remove(who) {
        let index = this.parts.indexOf(who)
        if (index === -1) {
            return
        }
        this.parts.splice(index, 1)
    }

    /**
     * Приглашен на копу?
     * @returns {Boolean} индекс
     */
    isAdded(who) {
        return this.parts.indexOf(who) !== -1
    }

    /**
     * Звать - не звать поменять наоборот
     * @param {Kopnik} who
     */
    toggle(who) {
        if (this.isAdded(who)) {
            this.remove(who)
        } else {
            this.add(who)
        }
    }

    inviteAll() {
        // alert("СОЗВАТЬ ВСЕХ НА КОПУ В РАЗРАБОТКЕ")
        this.parts.splice(0)
    }
}
