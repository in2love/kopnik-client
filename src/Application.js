import AsyncLock from 'async-lock'
import {AbstractSync, Kopnik, Kopa} from "./models";
import {KopnikApiError, KopnikError} from "./KopnikError";
import once from "./decorators/once";
import SquadAnalyzer from "./SquadAnalyzer";


export default class Application {
    constructor(logger) {
        this.logger = logger.getLogger('application')
        /**
         * Кэш моделей
         * @type {Array}
         */
        // пока не понятно зачем приложению кэш.
        // он вероятно будет сильно торможить приложение из-за реактивности App.$data в которой он находится
        // хотя он же не в App.$data а в App.$options !!
        // уже в дата потому тчо в options не реактивен
        //this.models = AbstractSync.cache

        /**
         * Идентификатор раздела
         * Равен названиею соответствующего компонента
         *
         * @type {string} Map | Profile | Thanks
         */
        this.section = Application.Section.Main
        this.sectionLocker = new AsyncLock

        /**
         * 0 потому что на undefined не срабатывает watcher,
         * а null занят под "пользователь не аутентифицирован"
         */
        this.user = undefined

        this.top20 = []
        this.squadAnalyzer = new SquadAnalyzer()
        this.kopa = new Kopa
        /**
         *
         * @type {Kopnik}
         */
        this.selected = null
        /**
         * Сообщения пользователю
         * @type {String[]}
         */
        this.infos= []
        /**
         * Обишки приложения
         * @type {Error[]}
         */
        this.errors= []
    }

    /**
     * Эксклюзивный доступ к section
     * @param {function} callback
     * @returns {Promise<any>}
     */
    lockSection(callback) {
        return this.sectionLocker.acquire('section', callback)
    }

    /**
     * Устанавливить секцию
     * Если секция не может быть установлена, возвращает иную
     *
     * @param {String} section
     * @returns {Promise<string>} Установленная секция
     */
    async setSection(section) {
        let result
        if (this.section === section) {
            return section
        }
        switch (section) {
            case Application.Section.Profile:
            case Application.Section.Witness:
            case Application.Section.Thanks:
                if (await this.resolveUser()) {
                    if (section===Application.Section.Witness && this.user.status!== Kopnik.Status.CONFIRMED){
                        result= await this.setSection(Application.Section.Main)
                    }
                    else {
                        result = section
                    }
                } else {
                    result = await this.setSection(Application.Section.Main)
                }
                break
            case Application.Section.Main:
                result = section
                break
            default:
                throw new KopnikError('Wrong route', 666)
        }
        // this.logger.info('move', this.section, '->', section)
        return this.section = result
    }

    /**
     * Исследует дружину
     * @param {Kopnik} kopnik
     * @returns {Promise<void>}
     */
    async analyzeSquad(kopnik) {
        await this.squadAnalyzer.analyze(kopnik)
    }

    async loadTop20() {
            this.top20 = await Promise.all([1, 2, 3, 4].map(each => Kopnik.get(each)))
            this.logger.warn('manual set foremans')
            Kopnik.getReference(1).rank = 4
            Kopnik.getReference(2).foreman = Kopnik.getReference(3)
            Kopnik.getReference(2).rank = 1
            Kopnik.getReference(3).foreman = Kopnik.getReference(1)
            Kopnik.getReference(3).rank = 3
            Kopnik.getReference(4).foreman = Kopnik.getReference(3)
            Kopnik.getReference(4).rank = 1

            Kopnik.getReference(1).ten = [Kopnik.getReference(3)]
            Kopnik.getReference(2).ten = []
            Kopnik.getReference(3).ten = [Kopnik.getReference(2), Kopnik.getReference(4)]
            Kopnik.getReference(4).ten = []
    }

    /**
     * Определиться с пользователем. Или он некий копник, или он null, то есть незарегистрирован на сервере
     * @returns {Promise<Kopnik>}
     */
    async resolveUser() {
        if (this.user === undefined) {
            await this.authenticate()
        }
        return this.user
    }

    getSharedState() {
        return {
            SECTION: this.section
        }
    }

    setState(state) {
        if (state.section) {
            this.section = state.section
        }
    }

    /**
     * Инициализирует ползователя при запуске приложения
     *
     * @returns {Promise<void>}
     */
    @once
    async authenticate() {
        try {
            let userAsPlain = (await Kopnik.fetchApi('get?ids='))[0]
            this.user = await Kopnik.get(userAsPlain.id)
            this.logger.info('user authenticated', this.user)
        } catch (err) {
            if ((err instanceof KopnikApiError) && err.message.match(/no.+aut/i)) {
                this.user = null
                this.logger.info('user not authenticated')
            } else {
                throw err
            }
        }
    }

    /**
     * Учстанавливает пользователем объект из VK.Auth callback
     *
     * @param vkUser
     * @returns
     */
    /*    async setVkUser(vkUser) {
            global.credentials = {
                uid: vkUser.uid,
                hash: vkUser.hash
            }
            this.user = await Kopnik.getByUid(vkUser.uid)
            if (!this.user) {
                this.user = new Kopnik
                this.user.merge({
                    firstName: vkUser.first_name,
                    lastName: vkUser.last_name,
                    photo: vkUser.photo,
                    smallPhoto: vkUser.photo_rec,
                })

                this.section = "Profile"
            }
            // this.user.uid= vkUser.uid
            // this.user.hash= vkUser.hash

            // localStorage.setItem("vkUser", JSON.stringify(vkUser))
        }
        */
}

Application.Section = {
    Main: 'Main',
    Profile: 'Profile',
    Witness: 'Witness',
    Thanks: 'Thanks',
}
