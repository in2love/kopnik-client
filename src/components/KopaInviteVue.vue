<template>
    <div class="d-flex flex-wrap justify-start align-start k-kopaInvite"
         :class="{'kopaInvite-long': $vuetify.breakpoint.smAndUp}">
        <v-badge v-for="eachAvatar of avatars" :key="eachAvatar.value.id" :content="eachAvatar.value.rank" bottom
                 color="orange" :offset-x="24" :offset-y="28">
            <v-avatar :value="eachAvatar.value"
                      :size="48" class="ml-1 mr-1 mb-2" :class="eachAvatar.className"
                      :title="eachAvatar.value.rankName"
                      @click="avatar_click(eachAvatar.value)" @dblclick="avatar_dblclick(eachAvatar.value)"
            >
                <v-img :src="eachAvatar.value.photo"></v-img>
            </v-avatar>
        </v-badge>
        <slot></slot>
    </div>
</template>
<script>
    import _ from 'lodash'
    import {bottle, container} from "../plugins/bottle";
    import i18n from "../plugins/i18n";
    import logger from "./mixin/logger";
    import {Kopa, Kopnik} from "../models";

    export default {
        name: "KopaInvite",
        mixins: [logger],
        components: {},

        data() {
            return {
                application: container.application,
            }
        },
        props: {
            value: {
                type: Kopa
            }
        },
        computed: {
            avatars() {
                const result = this.value.parts
                    .map(eachPart => {
                        return {
                            value: eachPart,
                            className: 'map_avatar' + (this.application.user === eachPart ? ' map_avatar-user' : '') + (this.application.selected === eachPart ? ' map_avatar-selected' : ''),
                        }
                    })
                // console.log(result)
                return result
            },
        },
        watch: {},
        methods: {
            avatar_click(event) {
                this.$emit('avatar_click', event)
                return false
            },
            avatar_dblclick(event) {
                this.$emit('avatar_dblclick', event)
                return false
            },
        },
        async created() {

        },
        async mounted() {

        }
    }
</script>
<style lang="scss">
    $avatar-size: 48px;
    $margin: 4px;
    $element-width: $margin+$avatar-size+$margin;

    .k-kopaInvite {
        width: $element-width*5;
        pointer-events: none;
        /*outline: solid 2px black;*/
    }

    .k-kopaInvite .v-avatar, .k-kopaInvite button {
        pointer-events: all;
    }

    .k-kopaInvite-long {
        width: $element-width*10;
    }

    .k-kopaInvite > .v-avatar {
        transition-property: box-shadow, transform, opacity;
    }
</style>
