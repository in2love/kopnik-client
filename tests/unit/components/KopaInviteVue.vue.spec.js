// import {vueI18n} from '../../vue-setup'
import VueI18n from 'vue-i18n'
import {mount} from '@vue/test-utils'

import KopaInviteVue from '../../../src/components/KopaInviteVue'
import {Kopnik, Kopa} from "../../../src/models";
import flushPromises from "flush-promises";
import {vuetify} from "../../test-setup";

describe('unit components KopaInviteVue.vue', () => {
    it('render', async () => {
        const kopnik2 = await Kopnik.get(2),
            kopa= new Kopa

        kopa.add(kopnik2)
        const wrapper = mount(KopaInviteVue, {
            vuetify,
            propsData: {value: kopa}
        })
        expect(wrapper.element).toMatchSnapshot()
    })
})
