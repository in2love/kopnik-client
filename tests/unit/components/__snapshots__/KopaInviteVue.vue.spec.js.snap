// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit components KopaInviteVue.vue render 1`] = `
<div
  class="d-flex flex-wrap justify-start align-start kopaInvite kopaInvite-long"
>
  <div
    class="v-avatar ml-1 mr-1 mb-2"
    style="height: 48px; min-width: 48px; width: 48px;"
    title="Баранов Алексей qwer"
    value="Kopnik {2:\\"Баранов Алексей qwer\\"}"
  >
    <div
      class="v-responsive v-image"
    >
      <div
        class="v-image__image v-image__image--preload v-image__image--cover"
        mode="in-out"
        name="fade-transition"
        style="background-position: center center;"
      />
      <div
        class="v-responsive__content"
      />
    </div>
  </div>
   
</div>
`;
